packages/devtools_app_shared/lib/src/service/service_manager.dart

```diff
@@ -521,32 +522,74 @@ class ServiceManager<T extends VmService> {
     // VM service connection spawned from `dart test` or `flutter test`).
     if (packageRootUriString?.endsWith('.dart') ?? false) {
       final rootLibrary = await _mainIsolateRootLibrary();
-      final testTargetFileUriString =
-          (rootLibrary?.dependencies ?? <LibraryDependency>[])
-              .firstWhereOrNull((dep) => dep.prefix == 'test')
-              ?.target
-              ?.uri;
-      if (testTargetFileUriString != null) {
-        _log.fine(
-          '[connectedAppPackageRoot] detected test library from root library '
-          'imports: $testTargetFileUriString',
-        );
-        packageRootUriString = await packageRootFromFileUriString(
-          testTargetFileUriString,
-          dtd: dtdManager.connection.value,
-        );
-        _log.fine(
-          '[connectedAppPackageRoot] package root for test target: '
-          '$packageRootUriString',
-        );
+      if (rootLibrary != null) {
+        packageRootUriString =
+            (await _lookupPackageConfigByEval(rootLibrary)) ??
+                // TODO(kenz): remove this fallback once all test bootstrap
+                // generators include the `packageConfigLocation` constant we
+                // can evaluate.
+                await _lookupTestLibraryByPrefix(rootLibrary, dtdManager);
```

Comment:
@jakemac53 @natebosch @bkonyi is this backwards compatibility required? DevTools is pinned to the Dart SDK version. Is `package:test` also? Or is it possible for a user to be on a new Dart SDK with these DevTools changes, but an old version of `package:test` that does not have the generated constant we are trying to eval?

Reply:
Yes, it's possible for a new SDK to be used with an old `package:test`. We should not rely on these always being present. We also should be resilient if the strings have the content `'null'` in case these code paths are used in places where `Isoalte.packageConfig` returns `null`.

Reply:
Sounds good. The current code covers that case, so we should be good to go. In the case you described where `Isolate.packageConfig` returns 'null`, would we still want to fallback to the case where we are checking the test import prefix? From the DevTools perspective, I don't mind more fallbacks that if it allows us to provide a more robustness experience to the user.

Reply:
> would we still want to fallback to the case where we are checking the test import prefix? From the DevTools perspective

I don't know of any concrete reasons for getting a `null`, so it's hard to say whether the normal fallback behavior would help or not. In any case I don't think it hurts to try a fallback approach for this situation. I mentioned it because the API is nullable and we are explicitly _not_ doing anything special with it there. If it would be useful to omit the variable entirely over writing the string 'null' we could do that too.

Reply:
Let's leave the variable even if it writes null. If the variable is missing, the evaluation would throw, so I think it is better to just rely on our string comparison to cover this case.